Input rewrite: Narr review

flTobi opened the pull request:
thanks

flTobi committed:
add input common changes

flTobi committed:
add common changes

flTobi committed:
Stash "DONE"

german77 commented on src/citra_qt/bootmanager.cpp:
```diff
@@ -209,13 +213,16 @@ GRenderWindow::GRenderWindow(QWidget* parent_, EmuThread* emu_thread, bool is_se
     layout->setContentsMargins(0, 0, 0, 0);
     setLayout(layout);
 
+    input_subsystem->Initialize();
```
Since citra has two render windows this get's called twice. So it's better to initialize it on main.

german77 commented on src/citra_qt/bootmanager.cpp:
```diff
@@ -287,75 +294,123 @@ void GRenderWindow::closeEvent(QCloseEvent* event) {
 void GRenderWindow::keyPressEvent(QKeyEvent* event) {
-    InputCommon::GetKeyboard()->PressKey(event->key());
+    /**
+     * This feature can be enhanced with the following functions, but they do not provide
```
This comment is useless here. You don't have external keyboard support on 3ds

german77 commented on src/citra_qt/bootmanager.cpp:
```diff
@@ -287,75 +294,123 @@ void GRenderWindow::closeEvent(QCloseEvent* event) {
 void GRenderWindow::keyReleaseEvent(QKeyEvent* event) {
-    InputCommon::GetKeyboard()->ReleaseKey(event->key());
+    /**
+     * This feature can be enhanced with the following functions, but they do not provide
```
Ditto

german77 commented on src/citra_qt/configuration/configure_motion_touch.cpp:
```diff
@@ -34,26 +37,27 @@ CalibrationConfigurationDialog::CalibrationConfigurationDialog(QWidget* parent,
-                break;
-            }
-            QMetaObject::invokeMethod(this, "UpdateLabelText", Q_ARG(QString, text));
+            QMetaObject::invokeMethod(this, [status, this] {
```
Yuzu supports multiple udp clients with motion on all of them at once. Manual mapping auto detects the input device based on shake input. Selecting Input device auto maps motion if controller supports it. 

Decide which one you want since seems this window is now half broken. Since it's using previous UI and the controller UI has no motion button. And you deleted a lot of functions related to detecting motion source.

german77 commented on src/citra_qt/configuration/configure_motion_touch.cpp:
```diff
@@ -320,7 +274,7 @@ void ConfigureMotionTouch::ShowUDPTestResult(bool result) {
 
 void ConfigureMotionTouch::OnConfigureTouchFromButton() {
-    ConfigureTouchFromButton dialog{this, touch_from_button_maps,
+    ConfigureTouchFromButton dialog{this, touch_from_button_maps, input_subsystem,
```
Touch provider and use button mapping settings are no longer necessary. Kraken can handle all of them at once without conflicting.

german77 commented on src/citra_qt/main.cpp:
```diff
@@ -287,7 +295,6 @@ void GMainWindow::InitializeWidgets() {
         }
     });
 
-    InputCommon::Init();
```
It's probably wise to initialize input here. before the GRenderWindows. Since otherwise it will be initialized twice or finalized if one of the GRenderWindows is closed.

german77 commented on src/citra_qt/bootmanager.cpp:
```diff
@@ -209,13 +213,16 @@ GRenderWindow::GRenderWindow(QWidget* parent_, EmuThread* emu_thread, bool is_se
 
-GRenderWindow::~GRenderWindow() = default;
+GRenderWindow::~GRenderWindow() {
+    input_subsystem->Shutdown();
```
Ditto. This needs to be done on main.

german77 commented on src/common/settings.h:
```diff
@@ -131,6 +131,21 @@ constexpr std::array<const char*, NumAnalogs> mapping = {{
 }};
 } // namespace NativeAnalog
 
+namespace NativeMotion {
```
We don't have multiple motion sensors. This is not needed

german77 commented on src/common/settings.h:
```diff
@@ -397,6 +412,7 @@ struct InputProfile {
     std::string name;
     std::array<std::string, NativeButton::NumButtons> buttons;
     std::array<std::string, NativeAnalog::NumAnalogs> analogs;
+    std::array<std::string, NativeMotion::NumMotions> motions;
```
Just `std::string motions;` is good enough here.

german77 commented on src/core/frontend/emu_window.cpp:
```diff
@@ -87,6 +51,19 @@ static bool IsWithinTouchscreen(const Layout::FramebufferLayout& layout, unsigne
+        static_cast<float>(framebuffer_layout.screen.bottom - framebuffer_layout.screen.top);*/
+
+    // return std::make_pair(x, y);
+    return std::make_pair(0, 0);
```
```suggestion
    const float x =
        static_cast<float>(framebuffer_x - framebuffer_layout.bottom_screen.left) /
                    static_cast<float>(framebuffer_layout.bottom_screen.right -
                                       framebuffer_layout.bottom_screen.left);
    const float y =
        static_cast<float>(framebuffer_y - framebuffer_layout.bottom_screen.top) /
                    static_cast<float>(framebuffer_layout.bottom_screen.bottom -
                                       framebuffer_layout.bottom_screen.top);

     return std::make_pair(x, y);
```
The code is mostly the same. But you refer to the bottom screen :P

german77 commented on src/core/hid/emulated_controller.cpp:
```diff
@@ -0,0 +1,510 @@
+    callback_list.erase(iterator);
+}
+
+} // namespace Core::HID
```
New line

german77 commented on src/core/hid/hid_types.h:
```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include "common/bit_field.h"
+#include "common/common_funcs.h"
+#include "common/common_types.h"
+#include "common/uuid.h"
```
Not needed anymore.

german77 commented on src/core/hle/service/hid/hid.cpp:
```diff
@@ -192,10 +148,10 @@ void Module::UpdatePadCallback(std::uintptr_t user_data, s64 cycles_late) {
     bool pressed = false;
     float x, y;
-    std::tie(x, y, pressed) = touch_device->GetStatus();
+    /* std::tie(x, y, pressed) = touch_device->GetStatus();
```
While kraken supports multitouch you can just use the first index here and ignore the rest or apply an average. The inputs are always ordered. This code is from emulated_console.cpp

german77 commented on src/core/hle/service/hid/hid.cpp:
```diff
@@ -233,7 +189,7 @@ void Module::UpdateAccelerometerCallback(std::uintptr_t user_data, s64 cycles_la
 
     Common::Vec3<float> accel;
-    std::tie(accel, std::ignore) = motion_device->GetStatus();
+    // std::tie(accel, std::ignore) = motion_device->GetStatus();
```
Kraken should have all this data along with orientation. You can just grab what you need here from the struct

german77 commented on src/core/hle/service/ir/extra_hid.cpp:
```diff
@@ -228,24 +226,21 @@ void ExtraHID::OnReceive(const std::vector<u8>& data) {
 
     float x, y;
-    std::tie(x, y) = c_stick->GetStatus();
+    // std::tie(x, y) = c_stick->GetStatus();
```
You can always request the shared pointer from the emulated controller from anywhere to read input data

german77 commented on src/core/hle/service/ir/ir_rst.cpp:
```diff
@@ -51,34 +50,22 @@ struct SharedMem {
-        Settings::values.current_input_profile.analogs[Settings::NativeAnalog::CStick]);
-}
-
 void IR_RST::UnloadInputDevices() {
```
You don't need to unload anything from core.  Only the front end should finalize input

german77 commented on src/citra_qt/bootmanager.cpp:
```diff
@@ -215,7 +219,7 @@ GRenderWindow::GRenderWindow(QWidget* parent_, EmuThread* emu_thread, bool is_se
 }
 
-GRenderWindow::~GRenderWindow() = default;
+GRenderWindow::~GRenderWindow() {}
```
GRenderWindow::~GRenderWindow() = default;

flTobi committed:
Stash "TODO"

flTobi committed:
narr don't look

flTobi committed:
fix input profiles

german77 committed:
yuzu: config: Avoid reading deleted object

german77 committed:
input_common: Simplify stick from button

flTobi committed:
Qt: Fix mouse scaling

Since yuzu-emu/yuzu#4949 any user that had a window scale different from 100% will have wrong mouse coordinates. This PR fixes the issue by removing the workaround for scaling since now it should be 1:1 at any scale.

This PR also splits mouse input into 3 different devices. A physical mouse, touch and finally one for controllers. This solves the issue Zeikken85 had where having emulated mouse enabled had a weird behavior on stick input. As a final protection against setting invalid configs a warning message will show up if emulated mouse is enabled.

Co-Authored-By: Narr the Reg <5944268+german77@users.noreply.github.com>

FearlessTobi commented:
:c
